### PR TITLE
Export dialog: add iOS to UI flow

### PR DESCRIPTION
**Pull request by cpirich**
* Expand Export dialog to handle the iOS UI flow (under a new, separate `exportIOS` experiment) - based on export spec: https://docs.google.com/document/d/1o-C8GOi45FCzUKPRnsyrNS8k10jUYOlrP2zIBPeO1X4/edit
* NOTE: for the iOS case, there are a few notable differences
  1. An Apple developer account is required
  1. An Expo.io account is required
  1. Code Studio will create a snack (project) on Expo.io under the Expo "code.org" account, but will not generate the IPA. The 10-15 minute wait will not occur within Code Studio. The student will transition to the Expo.io site, which will remix the snack under the student's Expo.io account, explain the export process, prompt them to enter their Apple credentials, and show progress UI while waiting until the IPA is ready.
* NOTE: the Expo `snack-web` changes are not yet available. They will be based on the approach in this branch (for anyone who is curious): https://github.com/expo/snack-web/compare/master...cpirich:cpirich/routes-for-save-and-publishipa

New screenshots (for review by @ryansloan):
Screen Shot 2019-08-05 at 10 55 47 AM
Screen Shot 2019-08-05 at 10 56 04 AM
Screen Shot 2019-08-05 at 10 56 11 AM
Screen Shot 2019-08-05 at 10 56 23 AM

### CODE REVIEW

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx [hunk 1]
```diff
@@ -337,6 +347,16 @@ class ExportDialog extends React.Component {
     }
   }
 
+  visitExpoSite() {
+    const {expoSnackId} = this.state;
+    if (!expoSnackId) {
+      return;
+    }
+    // TODO: use new URL format once snack-web has been updated for this flow
+    // TODO: pass iconUri and splashImageUri to expo.io
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/Dialog.jsx:356]: Are there work items for these?

File: apps/src/code-studio/components/ExportDialog/PlatformPage.jsx [hunk 2]
```diff
@@ -22,7 +24,27 @@ const styles = {
  * Platform Page in Export Dialog
  */
 export default class PlatformPage extends React.Component {
+  static propTypes = {
+    platform: PropTypes.oneOf(['ios', 'android']).isRequired,
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/PlatformPage.jsx]: Add 'ios' and 'android' to the applab/constants.js file.

File: apps/src/code-studio/components/ExportDialog/PublishIOSPage.jsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+          <p style={commonStyles.title}>Create iOS App Store Package</p>
+        </div>
+        <div style={commonStyles.section}>
+          <p style={commonStyles.text}>
+            An iOS App Store Package (IPA) is a package of code and other files
+            that can be installed as an app on an iOS device.
+          </p>
+          <p style={commonStyles.text}>
+            <b>Note: </b>You must have an Apple developer account to create an
+            IPA using our Code.org partner site, Expo.io.
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/PublishIOSPage.jsx:21]: Seems like this warning should be stronger @ryansloan - perhaps "REQUIRED: You must..."

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx (hunk 1 above)
Comment (jmkulwik) [reply apps/src/code-studio/components/ExportDialog/Dialog.jsx:356]: Also, for the link (below) should it redirect to the signup page? Right now, it sends me to a page with a bunch of code on it and I'm not sure what I'm supposed to do next.

**Comment by jmkulwik** [comment]
Is there a design spec for the iOS route yet? I don't quite understand what the final product should look like? How much of this is going to be done by Snack Expo?

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx [hunk 4]
```diff
@@ -446,13 +466,22 @@ class ExportDialog extends React.Component {
...
+          screen: platform === 'android' ? 'publishAndroid' : 'publishIOS'
+        });
         break;
-      case 'publish':
+      case 'publishAndroid':
         this.generateApkAsNeeded();
         this.setState({screen: 'generating'});
         break;
+      case 'publishIOS':
+        this.publishExpoExport();
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/Dialog.jsx:479]: Perhaps should be named `generateIpaAsNeeded()`

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx (hunk 1 above)
Comment (cpirich) [reply apps/src/code-studio/components/ExportDialog/Dialog.jsx:356]: Yes, the `snack-web` changes are not yet available (that is what the TODO notes are indicating).
Once those changes are available on `snack-web`, this code will be updated to navigate to something like `/publishipa/@:username/:projectName`, which will redirect to signup/login and then explain the IPA export process.

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx (hunk 4 above)
Comment (cpirich) [reply apps/src/code-studio/components/ExportDialog/Dialog.jsx:479]: This `publishExpoExport()` step is used by both iOS and Android.
For Android, we call `generateApkAsNeeded()`, which calls `publishAndGenerateApk()`, which does a 2 step process: first `publishExpoExport()`, then `expoGenerateApk()`
For iOS, we just do `publishExpoExport()`, and then we transition to Expo.io to continue.

**Comment by cpirich** [comment]
> Is there a design spec for the iOS route yet? I don't quite understand what the final product should look like? How much of this is going to be done by Snack Expo?

@ryansloan can comment on the status of the spec. At this point, it looks like I will be doing the UI work on the Expo side based on the branch I referenced above. The Expo team is still working this week on completing the IPA build process.

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx (hunk 4 above)
Comment (jmkulwik) [reply apps/src/code-studio/components/ExportDialog/Dialog.jsx:479]: 👍

**Comment by ryansloan** [comment]
@jmkulwik @cpirich There is not a spec on the Expo end of this yet. In order to do testing of the iOS export process we need to hook up our end of this. The expo team has prototyped an "export for app store" button on the snack page, but we need to work on the actual UI plan. This is behind an experiment flag separate from our Android release and we're not planning to release iOS at the same time.

**Review by jmkulwik — APPROVED** [review]

**Pull request merged**